File: backend-fiber/router/employee.go
package router

import (
	"backend-fiber/database"
	"database/sql"
	"fmt"

	"github.com/gofiber/fiber/v2"
)

type Employee struct{}

func NewEmployee() *Employee {
	return &Employee{}
}

// setup router
func (u *Employee) SetupRouter(app *fiber.App) {
	employee := app.Group("/api")
	employee.Get("/employees", u.employees)
}

// employees entity
type EmployeeEntity struct {
	EmpNo     int32  `json:"emp_no"`
	BirthDate string `json:"birth_date"`
	FirstName string `json:"first_name"`
	LastName  string `json:"last_name"`
	Gender    string `json:"gender"`
	HireDate  string `json:"hire_date"`
}

// employees
func (u *Employee) employees(c *fiber.Ctx) error {
	rows, err := database.Db.Query("SELECT emp_no, birth_date, first_name, last_name, gender, hire_date FROM employee LIMIT 20000,20")
	if err != nil {
		fmt.Printf("select employees error: %s\n", err.Error())
		return c.JSON(fiber.Map{
			"code": 1,
			"msg":  "select employees failed",
		})
	}

	employeeEntityArray := []EmployeeEntity{}

	for rows.Next() {
		var emp_no int32
		var first_name, last_name, gender string
		var birth_date, hire_date sql.NullTime

		rows.Scan(&emp_no, &birth_date, &first_name, &last_name, &gender, &hire_date)

		employeeEntity := EmployeeEntity{
			EmpNo:     emp_no,
			BirthDate: birth_date.Time.Format("2006-01-02"),
			FirstName: first_name,
			LastName:  last_name,
			Gender:    gender,
			HireDate:  hire_date.Time.Format("2006-01-02"),
		}

		employeeEntityArray = append(employeeEntityArray, employeeEntity)
	}

	return c.JSON(fiber.Map{
		"code": 0,
		"data": employeeEntityArray,
		"msg":  "success",
	})
}

File: backend-axum/src/router/employee.rs
use axum::{
    extract::{Query, State},
    http::Uri,
    Json,
    routing::get,
    Router
};
use serde::Serialize;
use serde_json::json;
use sqlx::MySqlPool;

use super::{parse_pagination, Pagination};

// init router
pub fn init_router() -> Router<MySqlPool> {
    let app = Router::new()
        .route("/employees", get(employees));
    Router::new()
        .nest("/api", app)
}

// employee entity
#[derive(Serialize, sqlx::FromRow)]
struct Employee {
    emp_no: i32,
    birth_date: chrono::NaiveDate,
    first_name: String,
    last_name: String,
    gender: String,
    hire_date: chrono::NaiveDate
}

// employees
async fn employees(
    State(pool): State<MySqlPool>,
    uri: Uri
) -> Json<serde_json::Value> {
    let params: Result<Query<Pagination>, _> = Query::try_from_uri(&uri);
    if params.is_err() {
        return Json(json!({
            "code": 1,
            "msg": "invalid parameters"
        }))
    }

    // parse pagination
    let (_, page_size, offset) = parse_pagination(params.unwrap().0);

    let result = sqlx::query_as::<_, Employee>("SELECT t1.emp_no, birth_date, first_name, last_name, gender, hire_date FROM employee t1 INNER JOIN (SELECT emp_no FROM employee ORDER BY emp_no LIMIT ?,?) t2 ON t1.emp_no = t2.emp_no")
        .bind(offset)
        .bind(page_size)
        .fetch_all(&pool)
        .await;

    match result {
        Ok(data) => {
            return Json(json!({
                "code": 0,
                "data": data,
                "msg": "success"
            }))
        },
        Err(err) => {
            tracing::error!("select employees error: {}", err.to_string());
            return Json(json!({
                "code": 1,
                "msg": "select employees failed"
            }))
        }
    }
}
